File: jar/src/main/java/jline/api/wf/WorkflowManager.java
/**
 * @file Workflow Management System
 *
 * @since LINE 3.0
 */
package jline.api.wf;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import jline.lang.Network;
import jline.solvers.NetworkSolver;
import jline.solvers.SolverOptions;
import jline.solvers.SolverResult;
import jline.util.matrix.Matrix;

/**
 * Main facade for workflow management and optimization in LINE.
 */
public class WorkflowManager {

    private final Network network;
    private final SolverOptions options;
    private final Wf_analyzer workflowAnalyzer;
    private final Wf_auto_integration autoIntegration;

    public WorkflowManager(Network network, SolverOptions options) {
        this.network = network;
        this.options = options;
        this.workflowAnalyzer = new Wf_analyzer(network);
        this.autoIntegration = new Wf_auto_integration(network, options);
    }

    public WorkflowManager(Network network) {
        this(network, new SolverOptions());
    }

    /** Comprehensive workflow analysis result. */
    public static final class WorkflowAnalysisResult {
        public final Wf_analyzer.WorkflowAnalysis patternAnalysis;
        public final Wf_auto_integration.ExtendedSolverRecommendation solverRecommendation;
        public final Map<String, Object> optimizationInsights;
        public final Map<String, Double> performanceMetrics;

        public WorkflowAnalysisResult(Wf_analyzer.WorkflowAnalysis patternAnalysis,
                                      Wf_auto_integration.ExtendedSolverRecommendation solverRecommendation,
                                      Map<String, Object> optimizationInsights,
                                      Map<String, Double> performanceMetrics) {
            this.patternAnalysis = patternAnalysis;
            this.solverRecommendation = solverRecommendation;
            this.optimizationInsights = optimizationInsights;
            this.performanceMetrics = performanceMetrics;
        }

        public Wf_analyzer.WorkflowAnalysis getPatternAnalysis() { return patternAnalysis; }
        public Wf_auto_integration.ExtendedSolverRecommendation getSolverRecommendation() { return solverRecommendation; }
        public Map<String, Object> getOptimizationInsights() { return optimizationInsights; }
        public Map<String, Double> getPerformanceMetrics() { return performanceMetrics; }
    }

    public WorkflowAnalysisResult analyzeWorkflow() {
        Wf_analyzer.WorkflowAnalysis patternAnalysis = workflowAnalyzer.analyzeWorkflow();
        Wf_auto_integration.ExtendedSolverRecommendation solverRecommendation =
                autoIntegration.recommendSolverWithWorkflowAnalysis();
        Map<String, Object> optimizationInsights = autoIntegration.getOptimizationInsights();
        Map<String, Double> performanceMetrics = calculatePerformanceMetrics(patternAnalysis, solverRecommendation);
        return new WorkflowAnalysisResult(patternAnalysis, solverRecommendation,
                optimizationInsights, performanceMetrics);
    }

    public NetworkSolver createOptimizedSolver() {
        return autoIntegration.createOptimalSolver();
    }

    public Wf_analyzer.DetectedPatterns getPatternAnalysis() {
        Wf_analyzer.WorkflowAnalysis analysis = workflowAnalyzer.analyzeWorkflow();
        return analysis.detectedPatterns;
    }

    public List<String> getOptimizationRecommendations() {
        Wf_analyzer.WorkflowAnalysis analysis = workflowAnalyzer.analyzeWorkflow();
        return workflowAnalyzer.getOptimizationRecommendations(analysis);
    }

    public Map<String, Object> generateComplexityReport() {
        Wf_analyzer.WorkflowAnalysis analysis = workflowAnalyzer.analyzeWorkflow();
        Map<String, Object> report = new HashMap<String, Object>();

        Object originalComplexityObj = analysis.statistics.get("originalComplexity");
        Object optimizedComplexityObj = analysis.statistics.get("optimizedComplexity");
        @SuppressWarnings("unchecked")
        Map<String, Object> originalComplexity = originalComplexityObj instanceof Map
                ? (Map<String, Object>) originalComplexityObj : null;
        @SuppressWarnings("unchecked")
        Map<String, Object> optimizedComplexity = optimizedComplexityObj instanceof Map
                ? (Map<String, Object>) optimizedComplexityObj : null;

        report.put("originalMetrics", originalComplexity != null ? originalComplexity : new HashMap<String, Object>());
        report.put("optimizedMetrics", optimizedComplexity != null ? optimizedComplexity : new HashMap<String, Object>());

        Wf_analyzer.DetectedPatterns patterns = analysis.detectedPatterns;
        Map<String, Object> patternComplexity = new HashMap<String, Object>();

        Map<String, Object> seqMap = new HashMap<String, Object>();
        seqMap.put("count", patterns.sequences.size());
        int seqTotalNodes = 0;
        int seqComplexity = 0;
        for (List<Integer> s : patterns.sequences) {
            seqTotalNodes += s.size();
            seqComplexity += s.size() * s.size();
        }
        seqMap.put("totalNodes", seqTotalNodes);
        seqMap.put("complexity", seqComplexity);
        patternComplexity.put("sequences", seqMap);

        Map<String, Object> parMap = new HashMap<String, Object>();
        parMap.put("count", patterns.parallels.size());
        int parTotalNodes = 0;
        int parComplexity = 0;
        for (List<Integer> p : patterns.parallels) {
            parTotalNodes += p.size();
            parComplexity += factorial(p.size());
        }
        parMap.put("totalNodes", parTotalNodes);
        parMap.put("complexity", parComplexity);
        patternComplexity.put("parallels", parMap);

        Map<String, Object> loopMap = new HashMap<String, Object>();
        loopMap.put("count", patterns.loops.size());
        loopMap.put("complexity", patterns.loops.size() * 10);
        patternComplexity.put("loops", loopMap);

        Map<String, Object> branchMap = new HashMap<String, Object>();
        branchMap.put("count", patterns.branches.size());
        int totalBranches = 0;
        int branchComplexity = 0;
        for (Wf_branch_detector.BranchPattern bp : patterns.branches) {
            totalBranches += bp.getBranchNodes().size();
            branchComplexity += bp.getBranchNodes().size();
        }
        branchMap.put("totalBranches", totalBranches);
        branchMap.put("complexity", branchComplexity);
        patternComplexity.put("branches", branchMap);

        report.put("patternComplexity", patternComplexity);

        int originalNodes = 0;
        int originalLinks = 0;
        if (originalComplexity != null) {
            Object n = originalComplexity.get("totalNodes");
            Object l = originalComplexity.get("totalLinks");
            if (n instanceof Number) originalNodes = ((Number) n).intValue();
            if (l instanceof Number) originalLinks = ((Number) l).intValue();
        }
        double complexityScore = calculateComplexityScore(originalNodes, originalLinks, patterns);
        report.put("overallComplexityScore", complexityScore);

        String level;
        if (complexityScore < 10) level = "LOW";
        else if (complexityScore < 50) level = "MEDIUM";
        else if (complexityScore < 100) level = "HIGH";
        else level = "VERY_HIGH";
        report.put("complexityLevel", level);

        return report;
    }

    public Map<String, Map<String, Object>> benchmarkSolvers(List<String> solvers) {
        Map<String, Map<String, Object>> results = new HashMap<String, Map<String, Object>>();
        for (String solverName : solvers) {
            try {
                NetworkSolver solver = createSolver(solverName);
                long startTime = System.currentTimeMillis();
                SolverResult solverResult = solver.getAvg();
                long endTime = System.currentTimeMillis();

                Map<String, Object> metrics = new HashMap<String, Object>();
                metrics.put("success", true);
                metrics.put("solveTime", endTime - startTime);
                metrics.put("hasResults", solverResult != null);

                if (solverResult != null && solverResult.QN != null) {
                    metrics.put("totalQueueLength", solverResult.QN.elementSum());
                    metrics.put("avgQueueLength", solverResult.QN.elementSum() / solverResult.QN.length());
                }

                results.put(solverName, metrics);
            } catch (Exception e) {
                Map<String, Object> metrics = new HashMap<String, Object>();
                metrics.put("success", false);
                metrics.put("error", e.getMessage() != null ? e.getMessage() : "Unknown error");
                results.put(solverName, metrics);
            }
        }
        return results;
    }

    public Map<String, Map<String, Object>> benchmarkSolvers() {
        return benchmarkSolvers(Arrays.asList("MVA", "NC", "SSA", "FLUID"));
    }

    public String exportAnalysis(String format) {
        WorkflowAnalysisResult analysis = analyzeWorkflow();
        String fmt = format.toUpperCase();
        if (fmt.equals("JSON")) return exportToJson(analysis);
        if (fmt.equals("CSV")) return exportToCsv(analysis);
        if (fmt.equals("SUMMARY")) return exportToSummary(analysis);
        throw new IllegalArgumentException("Unsupported export format: " + format);
    }

    public String exportAnalysis() {
        return exportAnalysis("SUMMARY");
    }

    public Map<String, Object> validateWorkflow() {
        Map<String, Object> validation = new HashMap<String, Object>();
        List<String> issues = new ArrayList<String>();
        try {
            if (network.getNodes().isEmpty()) {
                issues.add("Network has no nodes");
            }
            Wf_analyzer.WorkflowAnalysis analysis = workflowAnalyzer.analyzeWorkflow();
            if (!workflowAnalyzer.validateAnalysis(analysis)) {
                issues.add("Workflow analysis validation failed");
            }
            if (!autoIntegration.validateWorkflowEnhancement()) {
                issues.add("Workflow-enhanced solver selection validation failed");
            }
            validation.put("isValid", issues.isEmpty());
            validation.put("issues", issues);
            validation.put("nodeCount", network.getNodes().size());
            validation.put("hasRouting", !network.getLinkedRoutingMatrix().isEmpty());
        } catch (Exception e) {
            validation.put("isValid", false);
            List<String> errs = new ArrayList<String>();
            errs.add("Validation error: " + e.getMessage());
            validation.put("issues", errs);
        }
        return validation;
    }

    private Map<String, Double> calculatePerformanceMetrics(
            Wf_analyzer.WorkflowAnalysis analysis,
            Wf_auto_integration.ExtendedSolverRecommendation recommendation) {
        Map<String, Double> metrics = new HashMap<String, Double>();
        Object updateStatsObj = analysis.statistics.get("updateStats");
        @SuppressWarnings("unchecked")
        Map<String, Object> updateStats = updateStatsObj instanceof Map ? (Map<String, Object>) updateStatsObj : null;
        double reductionRatio = 0.0;
        if (updateStats != null) {
            Object rr = updateStats.get("reductionRatio");
            if (rr instanceof Number) reductionRatio = ((Number) rr).doubleValue();
        }
        metrics.put("complexityReduction", reductionRatio);
        metrics.put("solverConfidence", recommendation.confidence);

        Wf_analyzer.DetectedPatterns patterns = analysis.detectedPatterns;
        metrics.put("sequenceEfficiency", calculateSequenceEfficiency(patterns.sequences));
        metrics.put("parallelEfficiency", calculateParallelEfficiency(patterns.parallels));
        metrics.put("loopEfficiency", calculateLoopEfficiency(patterns.loops, analysis.originalWorkflow.linkMatrix));
        metrics.put("branchEfficiency", calculateBranchEfficiency(patterns.branches));
        return metrics;
    }

    private double calculateSequenceEfficiency(List<List<Integer>> sequences) {
        if (sequences.isEmpty()) return 1.0;
        int totalLength = 0;
        for (List<Integer> s : sequences) totalLength += s.size();
        double avgLength = (double) totalLength / sequences.size();
        return Math.min(1.0, avgLength / 5.0);
    }

    private double calculateParallelEfficiency(List<List<Integer>> parallels) {
        if (parallels.isEmpty()) return 1.0;
        int totalParallelism = 0;
        for (List<Integer> p : parallels) totalParallelism += p.size();
        double avgParallelism = (double) totalParallelism / parallels.size();
        return Math.max(0.1, 1.0 - (avgParallelism - 2.0) / 10.0);
    }

    private double calculateLoopEfficiency(List<Integer> loops, Matrix linkMatrix) {
        if (loops.isEmpty()) return 1.0;
        // Simplified: average probability of 0.5 per loop
        double avgProb = 0.5;
        return 1.0 - avgProb;
    }

    private double calculateBranchEfficiency(List<Wf_branch_detector.BranchPattern> branches) {
        if (branches.isEmpty()) return 1.0;
        double sum = 0.0;
        for (Wf_branch_detector.BranchPattern bp : branches) {
            Map<String, Double> diversity = Wf_branch_detector.calculateBranchDiversity(bp);
            Double e = diversity.get("normalizedEntropy");
            sum += e != null ? e.doubleValue() : 0.0;
        }
        return sum / branches.size();
    }

    private double calculateComplexityScore(int nodes, int links, Wf_analyzer.DetectedPatterns patterns) {
        double score = (double) nodes + links * 0.5;
        for (List<Integer> s : patterns.sequences) score += s.size() * s.size() * 0.1;
        for (List<Integer> p : patterns.parallels) score += factorial(p.size()) * 0.2;
        score += patterns.loops.size() * 10;
        for (Wf_branch_detector.BranchPattern bp : patterns.branches) score += bp.getBranchNodes().size() * 2;
        return score;
    }

    private static int factorial(int n) {
        if (n <= 1) return 1;
        return n * factorial(n - 1);
    }

    private NetworkSolver createSolver(String solverName) {
        String s = solverName.toUpperCase();
        if (s.equals("MVA")) return new jline.solvers.mva.SolverMVA(network, options);
        if (s.equals("NC")) return new jline.solvers.nc.SolverNC(network, options);
        if (s.equals("SSA")) return new jline.solvers.ssa.SolverSSA(network, options);
        if (s.equals("FLUID")) return new jline.solvers.fluid.SolverFluid(network, options);
        if (s.equals("JMT")) return new jline.solvers.wrappers.jmt.SolverJMT(network, options);
        if (s.equals("CTMC")) return new jline.solvers.ctmc.SolverCTMC(network, options);
        throw new IllegalArgumentException("Unknown solver: " + solverName);
    }

    private String exportToJson(WorkflowAnalysisResult analysis) {
        StringBuilder json = new StringBuilder();
        json.append("{\n");
        json.append("  \"solver_recommendation\": \"").append(analysis.solverRecommendation.recommendedSolver).append("\",\n");
        json.append("  \"confidence\": ").append(analysis.solverRecommendation.confidence).append(",\n");
        json.append("  \"patterns\": {\n");
        json.append("    \"sequences\": ").append(analysis.patternAnalysis.detectedPatterns.sequences.size()).append(",\n");
        json.append("    \"parallels\": ").append(analysis.patternAnalysis.detectedPatterns.parallels.size()).append(",\n");
        json.append("    \"loops\": ").append(analysis.patternAnalysis.detectedPatterns.loops.size()).append(",\n");
        json.append("    \"branches\": ").append(analysis.patternAnalysis.detectedPatterns.branches.size()).append("\n");
        json.append("  }\n");
        json.append("}");
        return json.toString();
    }

    private String exportToCsv(WorkflowAnalysisResult analysis) {
        StringBuilder csv = new StringBuilder();
        csv.append("Metric,Value\n");
        csv.append("Recommended Solver,").append(analysis.solverRecommendation.recommendedSolver).append("\n");
        csv.append("Confidence,").append(analysis.solverRecommendation.confidence).append("\n");
        csv.append("Sequences,").append(analysis.patternAnalysis.detectedPatterns.sequences.size()).append("\n");
        csv.append("Parallels,").append(analysis.patternAnalysis.detectedPatterns.parallels.size()).append("\n");
        csv.append("Loops,").append(analysis.patternAnalysis.detectedPatterns.loops.size()).append("\n");
        csv.append("Branches,").append(analysis.patternAnalysis.detectedPatterns.branches.size()).append("\n");
        return csv.toString();
    }

    private String exportToSummary(WorkflowAnalysisResult analysis) {
        StringBuilder summary = new StringBuilder();
        summary.append("=== Workflow Analysis Summary ===\n\n");
        summary.append("Recommended Solver: ").append(analysis.solverRecommendation.recommendedSolver).append("\n");
        summary.append("Confidence: ").append(String.format("%.2f", analysis.solverRecommendation.confidence)).append("\n\n");
        summary.append("Detected Patterns:\n");
        summary.append("- Sequences: ").append(analysis.patternAnalysis.detectedPatterns.sequences.size()).append("\n");
        summary.append("- Parallels: ").append(analysis.patternAnalysis.detectedPatterns.parallels.size()).append("\n");
        summary.append("- Loops: ").append(analysis.patternAnalysis.detectedPatterns.loops.size()).append("\n");
        summary.append("- Branches: ").append(analysis.patternAnalysis.detectedPatterns.branches.size()).append("\n\n");
        summary.append("Reasoning:\n");
        int idx = 0;
        for (String reason : analysis.solverRecommendation.reasoning) {
            summary.append(idx + 1).append(". ").append(reason).append("\n");
            idx++;
        }
        return summary.toString();
    }

    /** Quick analysis method for simple workflow inspection. */
    public static String quickAnalysis(Network network) {
        WorkflowManager manager = new WorkflowManager(network);
        return manager.exportAnalysis("SUMMARY");
    }

    /** Get optimal solver for a network without detailed analysis. */
    public static NetworkSolver getOptimalSolver(Network network, SolverOptions options) {
        WorkflowManager manager = new WorkflowManager(network, options);
        return manager.createOptimizedSolver();
    }

    public static NetworkSolver getOptimalSolver(Network network) {
        return getOptimalSolver(network, new SolverOptions());
    }
}

File: jar/src/main/kotlin/jline/api/wf/WorkflowManager.kt
/**
 * @file Workflow Management System
 * 
 * Provides workflow management capabilities for queueing network analysis,
 * integrating pattern detection and automatic workflow construction with
 * LINE solver framework for complex system modeling and optimization.
 * 
 * @since LINE 3.0
 */
package jline.api.wf

import jline.lang.Network
import jline.solvers.NetworkSolver
import jline.solvers.SolverOptions
import jline.solvers.SolverResult
import jline.solvers.fluid.SolverFluid
import jline.util.matrix.Matrix
import java.util.*

/**
 * Main facade for workflow management and optimization in LINE.
 * 
 * This class provides a high-level interface for workflow analysis,
 * pattern detection, optimization, and intelligent solver selection
 * based on the AUTO workflow analysis algorithms from the MDN toolbox.
 * 
 * Usage example:
 * ```kotlin
 * val manager = WorkflowManager(network)
 * val analysis = manager.analyzeWorkflow()
 * val solver = manager.createOptimizedSolver()
 * val results = solver.getAvg()
 * ```
 */
class WorkflowManager(
    private val network: Network,
    private val options: SolverOptions = SolverOptions()
) {
    
    private val workflowAnalyzer = Wf_analyzer(network)
    private val autoIntegration = Wf_auto_integration(network, options)
    
    /**
     * Comprehensive workflow analysis result.
     */
    data class WorkflowAnalysisResult(
        val patternAnalysis: Wf_analyzer.WorkflowAnalysis,
        val solverRecommendation: Wf_auto_integration.ExtendedSolverRecommendation,
        val optimizationInsights: Map<String, Any>,
        val performanceMetrics: Map<String, Double>
    )
    
    /**
     * Perform comprehensive workflow analysis.
     * 
     * @return Complete analysis including pattern detection, solver recommendation, and optimization insights
     */
    fun analyzeWorkflow(): WorkflowAnalysisResult {
        // Perform pattern analysis
        val patternAnalysis = workflowAnalyzer.analyzeWorkflow()
        
        // Get solver recommendation with workflow insights
        val solverRecommendation = autoIntegration.recommendSolverWithWorkflowAnalysis()
        
        // Get optimization insights
        val optimizationInsights = autoIntegration.getOptimizationInsights()
        
        // Calculate performance metrics
        val performanceMetrics = calculatePerformanceMetrics(patternAnalysis, solverRecommendation)
        
        return WorkflowAnalysisResult(
            patternAnalysis,
            solverRecommendation,
            optimizationInsights,
            performanceMetrics
        )
    }
    
    /**
     * Create an optimized solver based on workflow analysis.
     * 
     * @return NetworkSolver configured for optimal performance on this workflow
     */
    fun createOptimizedSolver(): NetworkSolver {
        return autoIntegration.createOptimalSolver()
    }
    
    /**
     * Get detailed pattern detection results.
     * 
     * @return Detected workflow patterns with statistics
     */
    fun getPatternAnalysis(): Wf_analyzer.DetectedPatterns {
        val analysis = workflowAnalyzer.analyzeWorkflow()
        return analysis.detectedPatterns
    }
    
    /**
     * Get workflow optimization recommendations.
     * 
     * @return List of actionable optimization recommendations
     */
    fun getOptimizationRecommendations(): List<String> {
        val analysis = workflowAnalyzer.analyzeWorkflow()
        return workflowAnalyzer.getOptimizationRecommendations(analysis)
    }
    
    /**
     * Generate a workflow complexity report.
     * 
     * @return Detailed complexity analysis
     */
    fun generateComplexityReport(): Map<String, Any> {
        val analysis = workflowAnalyzer.analyzeWorkflow()
        val report = HashMap<String, Any>()
        
        // Basic complexity metrics
        val originalComplexity = analysis.statistics["originalComplexity"] as? Map<*, *>
        val optimizedComplexity = analysis.statistics["optimizedComplexity"] as? Map<*, *>
        
        report["originalMetrics"] = originalComplexity ?: emptyMap<String, Any>()
        report["optimizedMetrics"] = optimizedComplexity ?: emptyMap<String, Any>()
        
        // Pattern complexity breakdown
        val patterns = analysis.detectedPatterns
        val patternComplexity = HashMap<String, Any>()
        
        patternComplexity["sequences"] = mapOf(
            "count" to patterns.sequences.size,
            "totalNodes" to patterns.sequences.sumOf { it.size },
            "complexity" to patterns.sequences.sumOf { it.size * it.size } // Quadratic complexity estimate
        )
        
        patternComplexity["parallels"] = mapOf(
            "count" to patterns.parallels.size,
            "totalNodes" to patterns.parallels.sumOf { it.size },
            "complexity" to patterns.parallels.sumOf { factorial(it.size) } // Exponential complexity for fork-join
        )
        
        patternComplexity["loops"] = mapOf(
            "count" to patterns.loops.size,
            "complexity" to patterns.loops.size * 10 // Loop complexity multiplier
        )
        
        patternComplexity["branches"] = mapOf(
            "count" to patterns.branches.size,
            "totalBranches" to patterns.branches.sumOf { it.branchNodes.size },
            "complexity" to patterns.branches.sumOf { it.branchNodes.size }
        )
        
        report["patternComplexity"] = patternComplexity
        
        // Overall complexity score
        val originalNodes = originalComplexity?.get("totalNodes") as? Int ?: 0
        val originalLinks = originalComplexity?.get("totalLinks") as? Int ?: 0
        val complexityScore = calculateComplexityScore(originalNodes, originalLinks, patterns)
        
        report["overallComplexityScore"] = complexityScore
        report["complexityLevel"] = when {
            complexityScore < 10 -> "LOW"
            complexityScore < 50 -> "MEDIUM"
            complexityScore < 100 -> "HIGH"
            else -> "VERY_HIGH"
        }
        
        return report
    }
    
    /**
     * Benchmark different solvers on this workflow.
     * 
     * @param solvers List of solver names to benchmark
     * @return Performance comparison results
     */
    fun benchmarkSolvers(solvers: List<String> = listOf("MVA", "NC", "SSA", "FLUID")): Map<String, Map<String, Any>> {
        val results = HashMap<String, Map<String, Any>>()
        
        for (solverName in solvers) {
            try {
                val solver = createSolver(solverName)
                val startTime = System.currentTimeMillis()
                
                // Attempt to solve
                val solverResult = solver.getAvg()
                val endTime = System.currentTimeMillis()
                
                val metrics = HashMap<String, Any>()
                metrics["success"] = true
                metrics["solveTime"] = endTime - startTime
                metrics["hasResults"] = solverResult != null
                
                if (solverResult != null && solverResult.QN != null) {
                    metrics["totalQueueLength"] = solverResult.QN.elementSum()
                    metrics["avgQueueLength"] = solverResult.QN.elementSum() / solverResult.QN.length()
                }
                
                results[solverName] = metrics
                
            } catch (e: Exception) {
                val metrics = HashMap<String, Any>()
                metrics["success"] = false
                metrics["error"] = e.message ?: "Unknown error"
                results[solverName] = metrics
            }
        }
        
        return results
    }
    
    /**
     * Export workflow analysis to different formats.
     * 
     * @param format Export format ("JSON", "CSV", "SUMMARY")
     * @return Formatted analysis data
     */
    fun exportAnalysis(format: String = "SUMMARY"): String {
        val analysis = analyzeWorkflow()
        
        return when (format.uppercase()) {
            "JSON" -> exportToJson(analysis)
            "CSV" -> exportToCsv(analysis)
            "SUMMARY" -> exportToSummary(analysis)
            else -> throw IllegalArgumentException("Unsupported export format: $format")
        }
    }
    
    /**
     * Validate workflow structure and analysis results.
     * 
     * @return Validation results with any issues found
     */
    fun validateWorkflow(): Map<String, Any> {
        val validation = HashMap<String, Any>()
        val issues = ArrayList<String>()
        
        try {
            // Check network structure
            if (network.nodes.isEmpty()) {
                issues.add("Network has no nodes")
            }
            
            // Validate workflow analysis
            val analysis = workflowAnalyzer.analyzeWorkflow()
            if (!workflowAnalyzer.validateAnalysis(analysis)) {
                issues.add("Workflow analysis validation failed")
            }
            
            // Validate solver integration
            if (!autoIntegration.validateWorkflowEnhancement()) {
                issues.add("Workflow-enhanced solver selection validation failed")
            }
            
            validation["isValid"] = issues.isEmpty()
            validation["issues"] = issues
            validation["nodeCount"] = network.nodes.size
            validation["hasRouting"] = network.getLinkedRoutingMatrix().isNotEmpty()
            
        } catch (e: Exception) {
            validation["isValid"] = false
            validation["issues"] = listOf("Validation error: ${e.message}")
        }
        
        return validation
    }
    
    // Private helper methods
    
    private fun calculatePerformanceMetrics(
        analysis: Wf_analyzer.WorkflowAnalysis,
        recommendation: Wf_auto_integration.ExtendedSolverRecommendation
    ): Map<String, Double> {
        val metrics = HashMap<String, Double>()
        
        // Complexity reduction metric
        val updateStats = analysis.statistics["updateStats"] as? Map<*, *>
        val reductionRatio = updateStats?.get("reductionRatio") as? Double ?: 0.0
        metrics["complexityReduction"] = reductionRatio
        
        // Solver confidence
        metrics["solverConfidence"] = recommendation.confidence
        
        // Pattern efficiency scores
        val patterns = analysis.detectedPatterns
        metrics["sequenceEfficiency"] = calculateSequenceEfficiency(patterns.sequences)
        metrics["parallelEfficiency"] = calculateParallelEfficiency(patterns.parallels)
        metrics["loopEfficiency"] = calculateLoopEfficiency(patterns.loops, analysis.originalWorkflow.linkMatrix)
        metrics["branchEfficiency"] = calculateBranchEfficiency(patterns.branches)
        
        return metrics
    }
    
    private fun calculateSequenceEfficiency(sequences: List<List<Int>>): Double {
        if (sequences.isEmpty()) return 1.0
        
        val totalLength = sequences.sumOf { it.size }
        val avgLength = totalLength.toDouble() / sequences.size
        
        // Longer sequences are more efficient to optimize
        return minOf(1.0, avgLength / 5.0)
    }
    
    private fun calculateParallelEfficiency(parallels: List<List<Int>>): Double {
        if (parallels.isEmpty()) return 1.0
        
        val totalParallelism = parallels.sumOf { it.size }
        val avgParallelism = totalParallelism.toDouble() / parallels.size
        
        // Higher parallelism can be more efficient but also more complex
        return maxOf(0.1, 1.0 - (avgParallelism - 2.0) / 10.0)
    }
    
    private fun calculateLoopEfficiency(loops: List<Int>, linkMatrix: Matrix): Double {
        if (loops.isEmpty()) return 1.0
        
        // Lower loop probabilities are more efficient
        val avgProb = loops.map { 
            // Simplified - would calculate actual loop probability
            0.5
        }.average()
        
        return 1.0 - avgProb
    }
    
    private fun calculateBranchEfficiency(branches: List<Wf_branch_detector.BranchPattern>): Double {
        if (branches.isEmpty()) return 1.0
        
        val avgEntropy = branches.map { 
            Wf_branch_detector.calculateBranchDiversity(it)["normalizedEntropy"] as? Double ?: 0.0
        }.average()
        
        // Higher entropy means more balanced branches, which is more efficient
        return avgEntropy
    }
    
    private fun calculateComplexityScore(
        nodes: Int, 
        links: Int, 
        patterns: Wf_analyzer.DetectedPatterns
    ): Double {
        var score = nodes.toDouble() + links * 0.5
        
        // Add pattern complexity
        score += patterns.sequences.sumOf { it.size * it.size } * 0.1
        score += patterns.parallels.sumOf { factorial(it.size) } * 0.2
        score += patterns.loops.size * 10
        score += patterns.branches.sumOf { it.branchNodes.size } * 2
        
        return score
    }
    
    private fun factorial(n: Int): Int {
        return if (n <= 1) 1 else n * factorial(n - 1)
    }
    
    private fun createSolver(solverName: String): NetworkSolver {
        return when (solverName.uppercase()) {
            "MVA" -> jline.solvers.mva.SolverMVA(network, options)
            "NC" -> jline.solvers.nc.SolverNC(network, options)
            "SSA" -> jline.solvers.ssa.SolverSSA(network, options)
            "FLUID" -> jline.solvers.fluid.SolverFluid(network, options)
            "JMT" -> jline.solvers.jmt.SolverJMT(network, options)
            "CTMC" -> jline.solvers.ctmc.SolverCTMC(network, options)
            else -> throw IllegalArgumentException("Unknown solver: $solverName")
        }
    }
    
    private fun exportToJson(analysis: WorkflowAnalysisResult): String {
        // Simplified JSON export - would use proper JSON library
        val json = StringBuilder()
        json.append("{\n")
        json.append("  \"solver_recommendation\": \"${analysis.solverRecommendation.recommendedSolver}\",\n")
        json.append("  \"confidence\": ${analysis.solverRecommendation.confidence},\n")
        json.append("  \"patterns\": {\n")
        json.append("    \"sequences\": ${analysis.patternAnalysis.detectedPatterns.sequences.size},\n")
        json.append("    \"parallels\": ${analysis.patternAnalysis.detectedPatterns.parallels.size},\n")
        json.append("    \"loops\": ${analysis.patternAnalysis.detectedPatterns.loops.size},\n")
        json.append("    \"branches\": ${analysis.patternAnalysis.detectedPatterns.branches.size}\n")
        json.append("  }\n")
        json.append("}")
        return json.toString()
    }
    
    private fun exportToCsv(analysis: WorkflowAnalysisResult): String {
        val csv = StringBuilder()
        csv.append("Metric,Value\n")
        csv.append("Recommended Solver,${analysis.solverRecommendation.recommendedSolver}\n")
        csv.append("Confidence,${analysis.solverRecommendation.confidence}\n")
        csv.append("Sequences,${analysis.patternAnalysis.detectedPatterns.sequences.size}\n")
        csv.append("Parallels,${analysis.patternAnalysis.detectedPatterns.parallels.size}\n")
        csv.append("Loops,${analysis.patternAnalysis.detectedPatterns.loops.size}\n")
        csv.append("Branches,${analysis.patternAnalysis.detectedPatterns.branches.size}\n")
        return csv.toString()
    }
    
    private fun exportToSummary(analysis: WorkflowAnalysisResult): String {
        val summary = StringBuilder()
        summary.append("=== Workflow Analysis Summary ===\n\n")
        
        summary.append("Recommended Solver: ${analysis.solverRecommendation.recommendedSolver}\n")
        summary.append("Confidence: ${String.format("%.2f", analysis.solverRecommendation.confidence)}\n\n")
        
        summary.append("Detected Patterns:\n")
        summary.append("- Sequences: ${analysis.patternAnalysis.detectedPatterns.sequences.size}\n")
        summary.append("- Parallels: ${analysis.patternAnalysis.detectedPatterns.parallels.size}\n")
        summary.append("- Loops: ${analysis.patternAnalysis.detectedPatterns.loops.size}\n")
        summary.append("- Branches: ${analysis.patternAnalysis.detectedPatterns.branches.size}\n\n")
        
        summary.append("Reasoning:\n")
        for ((index, reason) in analysis.solverRecommendation.reasoning.withIndex()) {
            summary.append("${index + 1}. $reason\n")
        }
        
        return summary.toString()
    }
    
    companion object {
        /**
         * Quick analysis method for simple workflow inspection.
         */
        @JvmStatic
        fun quickAnalysis(network: Network): String {
            val manager = WorkflowManager(network)
            return manager.exportAnalysis("SUMMARY")
        }
        
        /**
         * Get optimal solver for a network without detailed analysis.
         */
        @JvmStatic
        fun getOptimalSolver(network: Network, options: SolverOptions = SolverOptions()): NetworkSolver {
            val manager = WorkflowManager(network, options)
            return manager.createOptimizedSolver()
        }
    }
}